allennlp/predictors/predictor.py

```diff
@@ -58,6 +61,86 @@ def predict_json(self, inputs: JsonDict) -> JsonDict:
         instance = self._json_to_instance(inputs)
         return self.predict_instance(instance)
 
+    def inputs_to_labeled_instances(self, inputs: JsonDict) -> List[Instance]:
+        """
+        Converts incoming json to a :class:`~allennlp.data.instance.Instance`,
+        runs the model on the newly created instance, and adds labels to the
+        :class:`~allennlp.data.instance.Instance`s given by the model's output.
+        Returns
+        -------
+        List[instance]
+            A list of :class:`~allennlp.data.instance.Instance`s
+        """
+        instance = self._json_to_instance(inputs)
+        outputs = self._model.forward_on_instance(instance)
+        new_instances = self.predictions_to_labeled_instances(instance, outputs)
+        return new_instances
+
+    def get_gradients(self, instances: List[Instance]) -> Dict[str, np.ndarray]:
+        """
+        Gets the gradients of the loss with respect to the model inputs.
+
+        Parameters
+        ----------
+        instances: List[Instance]
+
+        Returns
+        -------
+        Dict[str, np.ndarray]
+            Dictionary of gradient entries for each input fed into the model.
+            The keys have the form ``{grad_input_1: ..., grad_input_2: ... }``
+            up to the number of inputs given.
+
+        Notes
+        -----
+        Takes a ``JsonDict`` representing the inputs of the model and converts
+        them to :class:`~allennlp.data.instance.Instance`s, sends these through
+        the model :func:`forward` function after registering hooks on the embedding
+        layer of the model. Calls :func:`backward` on the loss and then removes the
+        hooks.
+        """
+        self._register_hooks()
+
+        dataset = Batch(instances)
+        dataset.index_instances(self._model.vocab)
+        outputs = self._model.decode(self._model.forward(**dataset.as_tensor_dict()))
+
+        loss = outputs['loss']
+        self._model.zero_grad()
+        loss.backward()
+
+        for hook in self.hooks:
+            hook.remove()
+
+        grad_dict = dict()
+        for idx, grad in enumerate(self.extracted_grads):
+            key = 'grad_input_' + str(idx + 1)
+            grad_dict[key] = grad.squeeze_(0).detach().cpu().numpy()
+
+        return grad_dict, outputs
+
+    def _register_hooks(self):
+        """
+        Registers a backward hook on the
+        :class:`~allennlp.modules.text_field_embedder.basic_text_field_embbedder.BasicTextFieldEmbedder`
+        class. Used to save the gradients of the embeddings for use in get_gradients()
+
+        When there are multiple inputs (e.g., a passage and question), the hook
+        will be called multiple times. We append all the embeddings gradients
+        to a list.
+        """
+        self.extracted_grads = [] # pylint: disable=attribute-defined-outside-init
```

Comment:
This is another pylint warning that I would not ignore.  Just put these in `__init__`, with a brief comment about what they do.  Alternatively, you could do what you do in the interpreter and just pass in the extracted grads list, and return the list of hooks, so you don't need `self` at all.  That'd be cleaner.

I'd probably also name this something like `register_embedding_gradient_hooks`, to be more descriptive.  In a general `Predictor` class, it's less obvious what this kind of hook method is for.